File: frontend/BusinessFlow/src/app/modules/crm/services/client.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiService, PagedResponse } from '../../../core/services/api.service';
import { Client } from '../models/crm.model';

@Injectable({ providedIn: 'root' })
export class ClientService {
  private readonly endpoint = '/clients';

  constructor(private api: ApiService) {}

  list(page = 0, size = 20, status?: string, tagId?: number | null): Observable<PagedResponse<Client>> {
    const params: any = {};
    if (status) params.status = status;
    if (tagId) params.tagId = tagId;
    return this.api.getPaged<Client>(this.endpoint, page, size, Object.keys(params).length ? params : undefined);
  }

  // Lightweight, ungated - use this for pickers/dropdowns outside the Clients admin page.
  listActive(): Observable<Client[]> {
    return this.api.get<Client[]>(`${this.endpoint}/active`);
  }

  getById(id: number): Observable<Client> {
    return this.api.get<Client>(`${this.endpoint}/${id}`);
  }

  // Self-service (CLIENT role) - resolves the caller's own Client record server-side
  getMyProfile(): Observable<Client> {
    return this.api.get<Client>(`${this.endpoint}/me`);
  }

  updateMyProfile(payload: Partial<Client>): Observable<Client> {
    return this.api.patch<Client>(`${this.endpoint}/me`, payload);
  }

  // Backend CreateClientRequest also takes email + password to provision the portal user
  create(payload: any): Observable<Client> {
    return this.api.post<Client>(this.endpoint, payload);
  }

  update(id: number, payload: Partial<Client>): Observable<Client> {
    return this.api.patch<Client>(`${this.endpoint}/${id}`, payload);
  }

  /**
   * Creates a portal login for this client and emails a one-time set-password
   * link. No password is chosen here - the client sets their own, so nothing
   * has to be communicated out of band.
   *
   * Safe to call again for a client who already has a login: it just issues a
   * fresh link, which is what you want when the first one expired.
   */
  inviteToPortal(id: number): Observable<Client> {
    return this.api.post<Client>(`${this.endpoint}/${id}/invite-portal`, {});
  }

  delete(id: number): Observable<void> {
    return this.api.delete<void>(`${this.endpoint}/${id}`);
  }
}

File: backend/ZuhooCMS/src/main/java/com/zuhoocms/modules/crm/client/ClientService.java
package com.zuhoocms.modules.crm.client;

import com.zuhoocms.enums.ClientStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface ClientService {

    ClientResponse create(CreateClientRequest request);

    /** Public, unauthenticated self-registration - the client picks their own company. */
    ClientResponse registerPublic(PublicClientRegisterRequest request);

    ClientResponse getById(Long id);

    ClientResponse getMyProfile();

    ClientResponse updateMyProfile(UpdateMyClientProfileRequest request);

    Page<ClientResponse> listAll(ClientStatus status, Long tagId, Pageable pageable);

    /** Lightweight, ungated - the active-client picker used by Invoices, Payment Receipts, and Pipeline. */
    List<ClientResponse> listActive();

      ClientResponse update(Long id, UpdateClientRequest request);

    /**
     * Creates (or links) a CLIENT-role login for this client and emails a
     * one-time set-password link. No password is ever chosen by staff, so none
     * has to be communicated out of band.
     */
    ClientResponse inviteToPortal(Long id);

    void delete(Long id);

    long getClientCount();

    boolean isClient(Long userId);
}
